_posts/2023-01-12-websockets.md

```diff
@@ -0,0 +1,171 @@
+exchanged with the concurrent clients achieved by **Ktor**. Most importantly, it indicates the low impact of the observed fluctuations along the simulation, showing that both technologies ensure stability during the run. 
+
+The code snippets for **Ktor** and **Webflux** echo servers are listed below.
+
+<p style="margin : 0; padding-bottom:0;"> <em> Ktor WebSocket server.</em> </p>
+```kotlin
+fun Application.configureSockets() {
+    install(WebSockets) {
+        pingPeriod = Duration.ofSeconds(10)
+        timeout = Duration.ofSeconds(15)
+        maxFrameSize = Long.MAX_VALUE
+        masking = false
+    }
+    routing {
+        webSocket("/ktor") {
+            for (frame in incoming) {
+                frame as? Frame.Text ?: continue
+                val userMessage = frame.readText()
+                send(userMessage)
+            }
+        }
+    }
+}
+```
+
+
+<p style="margin : 0; padding-bottom:0;"> <em> WebFlux WebSocket server.</em> </p>
+```java
+@Component
+public class ReactiveServerWebSocketHandler implements WebSocketHandler {
+
+    @Override
+    public @NotNull
+    Mono<Void> handle(@NotNull WebSocketSession session) {
+        return session.send(session.receive()
+                .map(WebSocketMessage::getPayloadAsText)
+                .map(session::textMessage)
+        );
+    }
+
+    @Override
+    public @NotNull
+    List<String> getSubProtocols() {
+        return WebSocketHandler.super.getSubProtocols();
+    }
+}
+```
+
+
+After this digression, we can conclude that **Ktor** has performed slightly better, although we have observed a very 
+high-quality performance for **WebFlux**. 
+
+However, we should point out that the presented results rely on a single run, which means they lack statistical ensemble wideness. 
+In other words, to give a better outcome about the best-performing, we should carry out many runs and consider the average results from each of them.
+
+## A Didactical Digression
+It is interesting to carry out the same experiment by adopting a thread-per-request model, to observe the performance gap with the event-loop models presented so far. 
+We can build up the same echo server employing this time the standard WebSocket engine provided by Spring Boot. 
+Carrying out the same experiment as explained, we achieved the following results.
+
+<center>
+    <img src="{{ site.url }}{{ site.baseurl }}/assets/images/websockets/normal-results.png" width="1000">
+    <br>
+    <em> In the top panel the number of connected users during the run is presented. In the bottom panel, the comparison between the even-loop-based echo server performances is shown. Image produced using raw data from JMeter. </em>
+</center>
+
+In this case, the echo server can't handle more than roughly 100 concurrent connections, as indicated by the overall results parameters of the run. In particular, an error percentage of around 13% and a lower server throughput (16.5 sec) are detected. 
```

Comment:
In the previous example you put a table with the results, while here you write them inside the paragraph. I think that presenting a table again together with the image would make the results more readable.